File: WithCallback/app/src/main/java/com/ducluanxutrieu/tikiapp/ui/home/FlashDealAdapter.java
package com.ducluanxutrieu.tikiapp.ui.home;

import android.icu.text.NumberFormat;
import android.os.Build;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.ducluanxutrieu.tikiapp.R;
import com.ducluanxutrieu.tikiapp.data.models.FlashDealModel;
import com.ducluanxutrieu.tikiapp.databinding.ItemFlashDealBinding;
import com.ducluanxutrieu.tikiapp.utiu.Util;

import java.util.ArrayList;
import java.util.Locale;

public class FlashDealAdapter extends RecyclerView.Adapter<FlashDealAdapter.FlashDealViewHolder> {
    private final ArrayList<FlashDealModel> listFlashDeal = new ArrayList<>();

    void setData(ArrayList<FlashDealModel> list) {
        listFlashDeal.clear();
        listFlashDeal.addAll(list);
        notifyDataSetChanged();
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    @NonNull
    @Override
    public FlashDealViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ItemFlashDealBinding binding = DataBindingUtil.inflate(
                LayoutInflater.from(parent.getContext()),
        R.layout.item_flash_deal,
                parent,
                false
        );
        return new FlashDealViewHolder(binding);
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    @Override
    public void onBindViewHolder(@NonNull FlashDealViewHolder holder, int position) {
        holder.bindData(listFlashDeal.get(position));
    }

    @Override
    public int getItemCount() {
        return listFlashDeal.size();
    }


    @RequiresApi(api = Build.VERSION_CODES.N)
    public static class FlashDealViewHolder extends RecyclerView.ViewHolder {
        private final NumberFormat formatter = NumberFormat.getNumberInstance(new Locale("vi"));
        final ItemFlashDealBinding item;

        public FlashDealViewHolder(@NonNull ItemFlashDealBinding item) {
            super(item.getRoot());
            this.item = item;
        }

        void bindData(FlashDealModel flashDealModel) {
            Glide.with(item.getRoot()).load(flashDealModel.getProduct().getThumbnailUrl()).into(item.ivFlashDeal);
            String price = formatter.format(flashDealModel.getProduct().getPrice()) + " ₫";
            int percent = (int) (100 - flashDealModel.getProductProgress().getPercent());
            item.tvPrice.setText(price);
            item.pbFlashDeal.setProgress(percent);

            String discountPercent = "-" + flashDealModel.getDiscountPercent() + "%";
            item.tvDiscountPercent.setText(discountPercent);

            String sellStatus =
                Util.getString(R.string.sold) + " " + flashDealModel.getProductProgress().getQtyOrdered();

            if (percent > 50) {
                item.icHotDeal.setVisibility(View.VISIBLE);
            } else item.icHotDeal.setVisibility(View.GONE);

            if (percent < 5) {
                sellStatus = Util.getString(R.string.just_opened_for_sale);
            }

            if (percent < 20) {
                item.pbFlashDeal.setProgress(20);
            }
            item.tvSellStatus.setText(sellStatus);
        }
    }
}

File: app/src/main/java/com/example/tikiapp/ui/home/FlashDealAdapter.kt
package com.example.tikiapp.ui.home

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.example.tikiapp.R
import com.example.tikiapp.databinding.ItemFlashDealBinding
import com.example.tikiapp.data.models.FlashDealModel
import com.uit.party.util.StringUtil
import java.text.NumberFormat
import java.util.*
import kotlin.collections.ArrayList

class FlashDealAdapter : RecyclerView.Adapter<FlashDealAdapter.FlashDealViewHolder>() {
    private val listFlashDeal = ArrayList<FlashDealModel>()

    fun setData(list: ArrayList<FlashDealModel>) {
        listFlashDeal.clear()
        listFlashDeal.addAll(list)
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): FlashDealViewHolder {
        val binding = DataBindingUtil.inflate<ItemFlashDealBinding>(
            LayoutInflater.from(parent.context),
            R.layout.item_flash_deal,
            parent,
            false
        )
        return FlashDealViewHolder(binding)
    }

    override fun getItemCount(): Int {
        return listFlashDeal.size
    }

    override fun onBindViewHolder(holder: FlashDealViewHolder, position: Int) {
        holder.bindData(listFlashDeal[position])
    }

    class FlashDealViewHolder(private val item: ItemFlashDealBinding) :
        RecyclerView.ViewHolder(item.root) {
        private val formatter: NumberFormat = NumberFormat.getNumberInstance(Locale("vi"))

        fun bindData(flashDealModel: FlashDealModel) {
            Glide.with(item.root).load(flashDealModel.product.thumbnail_url).into(item.ivFlashDeal)
            val price = formatter.format(flashDealModel.product.price) + " ₫"
            val percent = 100 - flashDealModel.progress.percent
            item.tvPrice.text = price
            item.pbFlashDeal.progress = percent.toInt()

            val discountPercent = "-${flashDealModel.discount_percent}%"
            item.tvDiscountPercent.text = discountPercent

            var sellStatus =
                "${StringUtil.getString(R.string.sold)} ${flashDealModel.progress.qty_ordered}"

            if (percent > 50) {
                item.icHotDeal.visibility = View.VISIBLE
            } else item.icHotDeal.visibility = View.GONE

            if (percent < 5) {
                sellStatus = StringUtil.getString(R.string.just_opened_for_sale)
            }

            if (percent < 20) {
                item.pbFlashDeal.progress = 20
            }
            item.tvSellStatus.text = sellStatus
        }
    }
}
